### DISCUSSION TITLE
Suggested task for Outreachy Summer 22: 2

### DISCUSSION THREAD

**Discussion by pgleeson** [ModECI/MDF#225] [Ideas]
A suggested set of tasks for applicants to the Outreachy 2022 call.

1) Create a new account on OSBv2: https://www.v2.opensourcebrain.org 
2) Create a new [workspace](https://docs.opensourcebrain.org/OSBv2/Workspaces.html) from the [NeuroMatch Academy DL course](https://www.v2.opensourcebrain.org/repositories/20) repository 
3) Follow a number of the courses/tutorials there, running and saving the notebooks in your OSB workspace.
4) Create a short new notebook: “Two interesting concepts from Machine Learning”. What we're looking for here is your ability to explain some simple concepts in an educational Jupyter notebook. 
5) Share the workspace publicly on OSB
6) Email us a link to the workspace.

**Comment by somyagr** [comment]
Thanks for the task! I've created a new workspace and made it public, started week1-day1 tutorial and I'm very excited about the learning process, will keep updating you here.

**Reply by somyagr** [reply]
I am doing the exercises on Google Colab, can anyone tell me how to add the tutorial notebooks to my workspace? I was trying [Add resource -> Via URL] but it's raising error.
@pgleeson

**Reply by pgleeson** [reply]
@somyagr You could save the notebook created on Google collab to your local machine, and then try uploadign that to your workspace:
<img width="313" alt="Screenshot 2022-04-05 at 18 31 05" src="https://user-images.githubusercontent.com/1556687/161815710-e3dcf982-0601-4bfc-82ff-16de73af8bf5.png">

I've not tried that before, it would be interesting to see if it works...

**Reply by Fifi-Banjoko** [reply]
Okay, thanks.

**Reply by somyagr** [reply]
Thanks!!

**Comment by shivani6320** [comment]
Hello, I've created my workspace from NeuroMatch Academy DL Course & currently going through the tutorials. I'll soon create my ML Notebook on workspace and mail you the link. Thank you @pgleeson for introducing me to OSB :)✨

**Reply by Fifi-Banjoko** [reply]
Hello, please did you listen to the podcast? Is that something we should do or should we just skip to the tutorials?

**Reply by shivani6320** [reply]
No...I started with week 1 day 1 tutorial.

**Reply by Fifi-Banjoko** [reply]
Okay. I did listen to the podcasts tho. It's actually quite interesting.

**Reply by pgleeson** [reply]
Feel free to spend as much or as little time on the NeuroMatch Academy materials as you like. There is an assumption you will have some ML experince already so some of the material may not be relevant.

**Reply by Mercytopsy** [reply]
Okay, will do that

**Comment by Fifi-Banjoko** [comment]
Quick question, did anyone else have an issue importing torchvision into the tutorial notebook? I kept getting a 'No module named torchvision found' error then 'cannot import name 'torchvision' from 'torch'' error when I tried a few fixes. Which is strange because I've had torchvision for a while. I ran pip freeze to check for the packages installed and torchvision was not among them even though when I run pip freeze on my local jupyter notebook torchvision comes up and I can import it. I thought perhaps the tutorial notebook creates an environment while I use it and tried installing torchvision into it but it did not download. I opened it in Colab and it worked just fine. I've moved on from that tutorial, I just thought to ask.

**Reply by somyagr** [reply]
Yeah same with me, it wasted a lot of my time, then I switched to google colab

**Reply by Fifi-Banjoko** [reply]
Oh so it wasn't just me. I did waste some time too. When I did everything possible, I figured the problem might be from the notebook.

**Reply by sanjayankur31** [reply]
Where are you folks running into these issues? On OSBv2? `pytorch` should already be installed in the environment there:

```
pip list | grep torch
torch                         1.8.0
```

If this is on your local machine, remember to:

- install it in your virtualenv
- activate your virtualenv before trying to use it.

**Reply by Fifi-Banjoko** [reply]
Yes on OSB. Pytorch is installed, I ran grep torch and I got the same result. Other libraries seem to work fine, the problem seems to be for just torchvision so far. This is on the virtual environment with pytorch installed and the virtual environment activated.

**Reply by sanjayankur31** [reply]
It may be because of a conflict in the versions of torch. The latest version of `torchvision` that one gets using `pip install` is 0.12.0, which seems to want `torch==1.11.0` but the default torch version we have installed on OSB is 1.8.0 :thinking:

**Reply by Fifi-Banjoko** [reply]
Oh interesting. Is there a fix for this?

**Reply by sanjayankur31** [reply]
try: `pip install torchvision==0.9.0` which seems to be happy with torch 1.8.0? `from torch install torchvision` still doesn't work though but `import torchvision` does.

**Reply by pgleeson** [reply]
Thanks for reporting! Yes, torch 1.8.0 is installed, but torchvision wasn't by default. If you just type `pip install torchvision` in the Terminal (which is a prefectly reasonable thing to do...) it will get the latest torchvision (not too big), which requires torch 1.11 (quite big...), and the install fails during download of the file. 
This is a known issue on OSBv2: https://github.com/OpenSourceBrain/OSBv2/issues/298 

A quick fix is to use an earlier version of torchvision: `pip install torchvision==0.9.0` compatible with the existing torch.

Another option is to use --no-cache-dir, e.g. `pip install torch==1.11.0 --no-cache-dir`

**Reply by pgleeson** [reply]
Torch 1.8 was preinstalled since that was what MDF required (the version dependency was pinned [here](https://github.com/ModECI/MDF/blob/main/setup.cfg#L35)).

The longer term solution is to
1) Unpin the dependencies for MDF and make sure it works with the latest versions of torch etc (@davidt0x has an issue for this here: https://github.com/ModECI/MDF/issues/230)
2) When we know it works fine, install the latest torch and torchvision by default in the jupyterlab environment of OSBv2 (see for example: https://github.com/OpenSourceBrain/OSBv2/pull/299/files)

Apologies if you've lost time on this @Fifi-Banjoko @somyagr, but it's finding and reporting issues like this that help multiple repositories and initiatives improve and become easier to use for the next user.

**Reply by Fifi-Banjoko** [reply]
Oh that explains why it didn't download. I have installed torchvision-0.9.0 and torch-1.8.0. I can see that torch 1.11.0 and torchvision 0.12.0 has been uninstalled. I restarted the kernel and tried importing torchvison, I still get the 'no module named 'torchvision'' error. I will try the longer term solution.

I would've reported this sooner but I thought maybe I was missing something so I tried to fix it first.
Thank you @pgleeson @sanjayankur31

**Reply by pgleeson** [reply]
Ok @Fifi-Banjoko. Just to note that new packages you pip install don't (currently) persist between sessions you have the workspace opened in, i.e. they don't get installed on the persistent storage. 

What you could do, if there were a key set of packages you required, is make a file `install_my_deps.sh` in the /opt/workspace directory with the commands to install them all, and run this file at the start of any new session.

The main issue is that the environment for jupyterlab is the same every time you start it up, and we can't really have every possible package preinstalled for everyone. We're investigating some options to make this easier, e.g. saying what packages need to be installed for specific repositories, in a config file committed to github: https://github.com/OpenSourceBrain/OSBv2/issues/325

**Reply by Fifi-Banjoko** [reply]
I just restarted the workspace and I can import torchvision now without any issues @pgleeson @sanjayankur31 @somyagr. Thanks!

Also, I went to the workspace terminal and created a script using vim and added the dependencies as suggested. I can run this at the started of my sessions. 

Thank you for the suggestion @pgleeson.
![package script](https://user-images.githubusercontent.com/29006834/161833136-5877438c-e7cb-49d6-85a0-0de12d766975.JPG)

**Reply by Fifi-Banjoko** [reply]
Start*

**Reply by somyagr** [reply]
Thankyou so much @pgleeson @sanjayankur31 @Fifi-Banjoko for your time, now I'm able to run the notebook without any error.

**Comment by Fifi-Banjoko** [comment]
I've also noticed that the notebook kernel started dying pretty often since I started running neural networks tutorials. I keep getting this error- "The kernel for Neuromatch Academy Deep Learning (NMA-DL)/main/tutorials/W1D2_LinearDeepLearning/W1D2_Tutorial3.ipynb appears to have died. It will restart automatically". I'd then have to run all the import cells again which slows down work. 

I'm not sure if this is from my end or from OSB.
![Notebook Error 2](https://user-images.githubusercontent.com/29006834/161841181-f059c093-dff4-4b20-b26e-852afe3303ae.JPG)

**Reply by sanjayankur31** [reply]
thanks, how long have you had the jupyterhub open there? I think it's a feature of jupyterhub that it stops the kernel after a while. We'll also look at our cloud bits to see if there's anything there that may be causing it in the meantime.

**Reply by Fifi-Banjoko** [reply]
It was open for several hours yesterday when I had the issue but so far everything has been working fine today. Thanks.

**Comment by DeejaKamran** [comment]
Heyy @pgleeson 
Than you for the task! I have started reading the tutorials. 
Also just wanted to ask if I am supposed to create the notebook “Two interesting concepts from Machine Learning” on my workspace?

**Reply by sanjayankur31** [reply]
Yes, you can use the workspace there for all your work pretty much---that includes working with tutorial bits, and creating your own notebooks and so on.

**Reply by DeejaKamran** [reply]
Okay great! Thank you.

**Comment by 29riyasaxena** [comment]
> A suggested set of tasks for applicants to the Outreachy 2022 call.
> 
> 1. Create a new account on OSBv2: https://www.v2.opensourcebrain.org
> 2. Create a new [workspace](https://docs.opensourcebrain.org/OSBv2/Workspaces.html) from the [NeuroMatch Academy DL course](https://www.v2.opensourcebrain.org/repositories/20) repository
> 3. Follow a number of the courses/tutorials there, running and saving the notebooks in your OSB workspace.
> 4. Create a short new notebook: “Two interesting concepts from Machine Learning”. What we're looking for here is your ability to explain some simple concepts in an educational Jupyter notebook.
> 5. Share the workspace publicly on OSB
> 6. Email us a link to the workspace.

Hi! Can you please describe the exact steps to see tutorials. I created a new workspace from the NeuroMatch Academy DL course repository but not able to see those tutorials. @pgleeson @sanjayankur31

**Reply by pgleeson** [reply]
Can you specify what step you are havign problems with?

You should be able to browse in the file list to a specific tutorial directory and open the notebooks, e.g.

<img width="1679" alt="Screenshot 2022-04-08 at 16 53 42" src="https://user-images.githubusercontent.com/1556687/162478493-5219818c-89ee-42a7-b3fd-8dba37375fac.png">

**Reply by 29riyasaxena** [reply]
> Can you specify what step you are havign problems with?
> 
> You should be able to browse in the file list to a specific tutorial directory and open the notebooks, e.g.
> 
> <img alt="Screenshot 2022-04-08 at 16 53 42" width="1679" src="https://user-images.githubusercontent.com/1556687/162478493-5219818c-89ee-42a7-b3fd-8dba37375fac.png">

I was unable to load these notebooks, but now it's fine. I am done with the task, please check your mail.

**Comment by somyagr** [comment]
![Screenshot from 2022-04-10 11-01-31](https://user-images.githubusercontent.com/78006502/162602950-7792e4d3-3c06-419d-858d-2d283a81bc2c.png)
The kernel dies after running the 12th cell automatically in W1D4 tutorial 1. I've ran the notebook after restarting the kernel many times but still the error persists. I don't know the problem is with my system or the OBS. Mentors @pgleeson @sanjayankur31 please look into this.

**Reply by Fifi-Banjoko** [reply]
I had the same issue even in some previous notebooks. I figured the notebook was using parameters that takes a lot of computational power and my computer wasn't keeping up. I switched to Colab, you can do that too.

**Reply by Fifi-Banjoko** [reply]
Colab seems to handle memory intensive operations better.

**Reply by somyagr** [reply]
Okay thankyou so much!

**Reply by Fifi-Banjoko** [reply]
You're welcome.

**Reply by pgleeson** [reply]
Apologies for this... We've not tried too many heavy duty ML notebooks on OSBv2 yet... The virtual nodes we pay for for running our jupyterlab instance are fairly basic, obviously the Google Collab ones have more resources. 

We'll look into this, but I doubt it will be resolved in the short term, so please do move to Google Collab if you want to continue with the NMA tutorials, after you have a feel for running and creating notebooks on OSB2.

**Reply by Fifi-Banjoko** [reply]
Oh okay. I still prefer the OSB notebooks though. I try them first before switching to Collab if I have to.

**Reply by pgleeson** [reply]
It's nice to hear we're doing something better than Google @Fifi-Banjoko!

**Reply by Fifi-Banjoko** [reply]
Haha yes. Its a really good platform.

**Reply by somyagr** [reply]
True @Fifi-Banjoko  , I really liked the OSB workspace! It is easy to use, and after working on it since a couple of weeks, I think I'm familiar with it and gives better user experience than most of the platforms. You guys are doing great @pgleeson .

**Reply by Fifi-Banjoko** [reply]
Yes I agree.

**Comment by 29riyasaxena** [comment]
Hi! I have done this task. Which date should I mention in the section "Date contribution was accepted or merged"? @pgleeson @sanjayankur31

**Comment by somyagr** [comment]
Hi, @pgleeson, while recording the contributions, do I have to insert the link of my OSB workspace in the Contribution URL section?

**Reply by Fifi-Banjoko** [reply]
That's what I did but I'm also waiting for a response to this.

**Reply by pgleeson** [reply]
Yes, you can use the url for the public workspace for the contribution url, but I would add in the description the location of the notebook which relates to the task

**Reply by somyagr** [reply]
Okay, thanks for the information.

**Reply by Fifi-Banjoko** [reply]
Thank you @pgleeson.
